scipy/optimize/_qap.py

```diff
@@ -0,0 +1,780 @@
+    B22 = B[n_seeds:, n_seeds:]
+
+    # setting initialization matrix
+    if isinstance(init_J, str) and init_J == 'barycenter':
+        J = np.ones((n_unseed, n_unseed)) / float(n_unseed)
+    else:
+        _check_init_input(init_J, n_unseed)
+        J = init_J
+
+    total_iter = 0
+    for i in range(init_k):
+        if init_weight is not None:
+            # generate a nxn matrix where each entry is a random number [0, 1]
+            # would use rand, but Generators don't have it
+            # would use random, but old mtrand.RandomStates don't have it
+            K = rng.uniform(size=(n_unseed, n_unseed))
+            # Sinkhorn balancing
+            K = _doubly_stochastic(K)
+            # initialize J, a doubly stochastic barycenter
+            P = J * init_weight + (1 - init_weight) * K
+        else:
+            P = J
+        const_sum = A21 @ B21.T + A12.T @ B12
+        grad_P = np.inf  # gradient of P
+        n_iter = 0  # number of FW iterations
+
+        # OPTIMIZATION WHILE LOOP BEGINS
+        while grad_P > eps and n_iter < maxiter:
+            # computing the gradient of f(P) = -tr(APB^tP^t)
+            delta_f = (const_sum + A22 @ P @ B22.T + A22.T @ P @ B22)
+            # run hungarian algorithm on gradient(f(P))
+            rows, cols = linear_sum_assignment(obj_func_scalar * delta_f)
+            Q = np.zeros((n_unseed, n_unseed))
+            Q[rows, cols] = 1  # initialize search direction matrix Q
+
+            def f(x):  # computing the original optimization function
+                xP1xQ = x * P + (1 - x) * Q
+                # Sums below are np.trace(A11.T @ B11)
+                # + np.trace(xP1xQ.T @ A21 @ B21.T)
+                # + np.trace(xP1xQ.T @ A12.T @ B12)
+                # + np.trace(A22.T @ xP1xQ @ B22 @ xP1xQ.T)
+                # This is more efficient, but can we do even better?
+                return obj_func_scalar * (
+                    (A11 * B11).sum()
+                    + (xP1xQ.T @ A21 * B21).sum()
+                    + (xP1xQ.T @ A12.T * B12.T).sum()
+                    + ((xP1xQ.T @ A22) * (B22 @ xP1xQ.T)).sum()
+                )
+
+            # computing the step size
+            alpha = minimize_scalar(f, bounds=(0, 1), method="bounded").x
+            P_i1 = alpha * P + (1 - alpha) * Q  # Update P
+            grad_P = np.linalg.norm(P - P_i1)
+            P = P_i1
+            n_iter += 1
+        # end of FW optimization loop
+
+        # Project onto the set of permutation matrices
+        row, col = linear_sum_assignment(-P)
+        perm_inds_new = np.concatenate((np.arange(n_seeds), col + n_seeds))
+
+        score = _calc_score(A, B, perm_inds_new)
+
+        # minimizing
+        if obj_func_scalar * score < obj_func_scalar * best_score:
+            best_score = score
+            perm_inds = np.zeros(n, dtype=int)
+            perm_inds[permutation_cost] = permutation_dist[perm_inds_new]
+            total_iter = n_iter
```

Comment:
Why is this only incremented if the score is better?
This is different behaviour to e.g. `optimize.fmin`.